File: Sorting/Quick Sort/QuickSort.py
# Partition Function
# Takes in an array and the left and right index
# Returns the position of the pivot
def partition(arr,l,r):

    # Pivot is the last element of the list
    pivot = arr[r]
    # i is the index of the first element of the list
    i = l-1

    # Looping through the list
    for j in range(l,r):
        # If the current element is less than the pivot
        # Then swap the current element with the element at index i
        if arr[j] <= pivot:
            i = i + 1
            arr[i],arr[j] = arr[j],arr[i]
    # Swap the pivot with the element at index i+1
    arr[i+1],arr[r] = arr[r],arr[i+1]
    return i+1

# Quicksort Function
# Takes in an array and the left and right index
# Returns the sorted array
def quicksort(arr,l,r):

    # Base case : if the left index is greater than the right index, then return
    if l >= r:
        return

    # Calling the partition function to get the position of the pivot
    p = partition(arr,l,r)

    # Recursively calling quicksort function on the list before the pivot
    quicksort(arr,l,p-1)
    # Recursively calling quicksort function on the list after the pivot
    quicksort(arr,p+1,r)


# Driver Code
lst = [2,4,1,8,5,7,6,3]
quicksort(lst,0,len(lst)-1)
print(lst)

# Code contributed by Sayan Maiti

File: Sorting/Quick Sort/QuickSort.cs
// C# program to illustrate
// Randomised Quick sort
using System;
class RandomizedQsort
{	

/* This function takes last element as pivot,
	places the pivot element at its correct
	position in sorted array, and places all
	smaller (smaller than pivot) to left of
	pivot and all greater elements to right
	of pivot */
static int partition(int[] arr, int low, int high)
{

	// pivot is chosen randomly
	random(arr, low, high);
	int pivot = arr[high];

	int i = (low-1); // index of smaller element
	for (int j = low; j < high; j++)
	{

	// If current element is smaller than or
	// equal to pivot
	if (arr[j] < pivot)
	{
		i++;

		// swap arr[i] and arr[j]
		int tempp = arr[i];
		arr[i] = arr[j];
		arr[j] = tempp;
	}
	}

	// swap arr[i+1] and arr[high] (or pivot)
	int tempp2 = arr[i + 1];
	arr[i + 1] = arr[high];
	arr[high] = tempp2;

	return i + 1;
}

// This Function helps in calculating
// random numbers between low(inclusive)
// and high(inclusive)
static int random(int[] arr, int low, int high)
{

	Random rand = new Random();
	int pivot = rand.Next() % (high - low) + low;

	int tempp1 = arr[pivot];
	arr[pivot] = arr[high];
	arr[high] = tempp1;

	return partition(arr, low, high);
}

/* The main function that implements Quicksort()
	arr[] --> Array to be sorted,
	low --> Starting index,
	high --> Ending index */
static void sort(int[] arr, int low, int high)
{
	if (low < high)
	{
	/* pi is partitioning index, arr[pi] is
			now at right place */
	int pi = partition(arr, low, high);

	// Recursively sort elements before
	// partition and after partition
	sort(arr, low, pi - 1);
	sort(arr, pi + 1, high);
	}
}

/* A utility function to print array of size n */
static void printArray(int[] arr)
{
	int n = arr.Length;
	for (int i = 0; i < n; ++i)
	Console.Write(arr[i] + " ");
	Console.WriteLine();
}

// Driver Code
static public void Main ()
{
	int[] arr = {10, 7, 8, 9, 1, 5};
	int n = arr.Length;

	sort(arr, 0, n-1);

	Console.WriteLine("sorted array");
	printArray(arr);
}
}

// This code is contributed by shubhamsingh10
